### ISSUE TITLE
Function.isCallable() / Function.isConstructor() example matrix

### ISSUE THREAD

**Issue by aedart** [caitp/TC39-Proposals#10]
Hi

Would it be possible to provide an example matrix with various argument types and the expected result of `isCallable()` and `isConstructor()`?

Currently, I'm looking at [Core-JS' implementation](https://github.com/zloirock/core-js#function-iscallable-isconstructor-) of this proposal and I'm wondering how built-in classes, e.g. `Array`, `Proxy`...etc, are expected to behave when tested via these methods.

**Comment by caitp** [comment]
It will take some work to put together a complete matrix, I don't know when I'll have the time to complete this. But my vision overall is:

Refers to isCallable() specifically:
- Most values with a [[Call]] internal method would return true
- Class constructors would return false
- A bound function would produce a value consistent with the target of the bound function
- Returns true for Proxy objects which have an "apply" trap, or whose target has a [[Call]] method that does not statically throw (e.g, returns false for class constructors or builtins which emulate that behaviour)
- Anything without a [[Call]] internal method, would return false (this is the case for most values, such as numbers, strings, most objects)

Similar logic would apply to isConstructor. If I get some time, and someone is interested in championing this in tc39, I'm happy to draft a matrix to help out

**Comment by aedart** [comment]
The reason why I'm asking about this, is to determine whether the implementation by Core-JS actually satisfies your vision, or if it is lacking. I'm particularly concerned with the about the ["isClassConstructor"](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/esnext.function.is-callable.js#L14) part in Core-JS (_not intended as criticism of the author_).

I guess what I'm trying to ask for, is if this proposal can be polyfilled. If so, perhaps that could be a place to start and continue to ask tc39 champions for help (_sorry, I'm not much help here because I'm not skilled enough_).

In any case, I do hope that this proposal gets through. JavaScript developers could really benefit from this type of "simple" functionality.

**Comment by caitp** [comment]
I believe this would have false negatives, in the case where a class constructor is bound. OTOH, a bound class constructor should fail the toString test, and the property descriptor test

I may be wrong about that, ill take a look not on mobile in a bit to be sure. Probably does not have false positives, though

**Comment by caitp** [comment]
Yeah,

```
class C {
  constructor() {
    console.log("C construct");
  }
}

let ctor = C.prototype.constructor.bind(null);

ctor(); // throws TypeError as it should
new ctor(); // returns instance of the class, invokes ctor
`${ctor}` // "function() {\n    [native code]\n}" 
```

(https://jsfiddle.net/utm79h2e/)

**Comment by zloirock** [comment]
Yep, AFAIK we have no way to detect bound classes. [So it's marked `sham: true`.](https://github.com/zloirock/core-js/blob/11f4ec8bb27f0f5513b6c61a68b8623c64f5a1b2/packages/core-js/modules/esnext.function.is-callable.js#L26)

**Comment by caitp** [comment]
yeah, it's not fully polyfillable, but it gets pretty close.

**Comment by zloirock** [comment]
BTW I'm planning to remove proposals that have not progressed for a long time from `core-js@4`. Do you have any plans to advance it?

**Comment by caitp** [comment]
It needs a champion to really get it anywhere, I can talk to my colleagues about raising it in committee, but it's not a super high priority atm

**Comment by aedart** [comment]
Thanks for your input. I really hope that this proposal will get through.

**Comment by icecream17** [comment]
I think I made an (almost perfect) polyfill for this:

```js
const isCallable = (val) => {
   try {
      [].filter(val)
      return true
   } catch (err) {
      if (err instanceof TypeError) return false

      // Impossible unless there's some KeyboardInterrupt equivalent
      throw err
   }
}
class _Ignore { }
const isConstructor = (arg) => {
   try {
      Reflect.construct(_Ignore, [], arg)
      return true
   } catch (err) {
      if (err instanceof TypeError) return false

      // Impossible unless there's some KeyboardInterrupt equivalent
      throw err
   }
}
```

Edit: Oh, callable means that also it is not a class constructor, instead of the internal IsCallable. Which is different from just isConstructor... I'll hide my comment
